lib/wibox/widget/base.lua

```diff
@@ -700,6 +736,20 @@ function base.make_widget(proxy, widget_name, args)
     mt.__tostring = function()
         return string.format("%s (%s)", ret.widget_name, orig_string)
     end
+
+    -- Even when properties are disabled, buttons is required for backward
+    -- compatibility.
+    --TODO v6 Remove this
```

Comment:
Actual v6 or typo v5?

Reply:
Actual v6. This API is very hard to both modernize and keep backward compatibility. The goal of this branch is to have a real read/write API for input handling. The old one is write only. Once you set the keys/buttons, they lose their context and you have to reverse engineer everything to manipulate them later on. It will need long term code to support the old deprecated API. Both for 4.x (not officially deprecated yet) and 5.x (deprecated but supported) so we can remove it from 6.0 (deprecated and free to break if the maintenance cost reduction is worth it, which in this case it totally will be).

/offtopic

This isn't really a pull request anyway. It's more like a branch with many, many small, long term, independent, commits that can be reviewed individually (as long as the order remain the same). Once I will be done with this, there will be a real API to manipulate input events. But given that I want to create something that's easier to review than my usual multi-hundred lines commits, its going to take some iterations of changes within this branch. I add new APIs in some commit, then "slowly"  integrate them, then tests them, then replace the usage of the old one, then deprecate it. All in different commits. It will most probably have >100 commits, but commits that I hope can be reviewed/merged with less pain than my usual work.

psychon doesn't have time to review things right now (which is 100% ok, I am in the same situation most of the year), it's better to spend my AwesomeWM time on this than on the notification PR series because it is stuck as long as the open ones are not merged. That, in turn requires both of us active at the same time, which hasn't been happening a lot this year I know you and others want the notification API and don't care about much about API standardization. The problem is that stars will have to align to be able to merge that work and it isn't the case right now.

Reply:
> Actual v6. This API is very hard to both modernize and keep backward compatibility. The goal of this branch is to have a real read/write API for input handling. The old one is write only. Once you set the keys/buttons, they lose their context and you have to reverse engineer everything to manipulate them later on. It will need long term code to support the old deprecated API. Both for 4.x (not officially deprecated yet) and 5.x (deprecated but supported) so we can remove it from 6.0 (deprecated and free to break if the maintenance cost reduction is worth it, which in this case it totally will be).

Got it! Just curious when looking over it.

> psychon doesn't have time to review things right now (which is 100% ok, I am in the same situation most of the year), it's better to spend my AwesomeWM time on this than on the notification PR series because it is stuck as long as the open ones are not merged. That, in turn requires both of us active at the same time, which hasn't been happening a lot this year

Yeah, I've been focused on other things, too. Haven't touched awesome source in a while.

> I know you and others want the notification API and don't care about much about API standardization.

On the contrary! I very much am supportive for API standardization. I'm just not as versed in the core source as y'all are. I try where I can. :)